Feature: exclude bridges/tunnels/tolls

azarz opened the pull request:
# Issue

Issue #3587
I have the specific need to be able to exclude bridges, tunnels and/or tolls when computing routes or isochrones (https://www.geoportail.gouv.fr/carte). The fact that no path may be found when those options are on is a feature: I need to know whether a place can be reached without taking any tunnels for example.

This PR introduces 3 costing options, `exclude_bridges`, `exclude_tunnels` and `exclude_tolls`. If set to 1, the path or isochrone will be able to start or end on bridges/tunnels/tolls, but will exclude them in any routing in-between. If set to 1, it is highly plausible that no path will be found.

The default server behavior is to disallow these options: if they are set to 1, a 400 error code will be returned to the user. To allow those options on a server, the parameter `service_limits.status.allow_hard_exclusions` must be set to `true` in the JSON server configuration.

## Tasklist

 - [x] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [x] Update the docs with any new request parameters or changes to behavior described
 - [x] Update the [changelog](CHANGELOG.md)
 - [ ] If you made changes to the lua files, update the [taginfo](taginfo.json) too.

## Requirements / Relations
Based on  #3240

dnesbitt61 commented:
I do not think there should be a hard avoid for bridges. Even simple overpasses on US Interstate highways are tagged as bridges in OSM so any exclude bridges route in US would likely never take and interstate highway and likely would fail or provide strange directions for many routes. My opinion is this is fine for your custom fork but is likely to lead to issues and complaints were this to be part of the mainstream Valhalla service.

nilsnolde commented:
I’m not very opinionated for these options, IMHO who doesn’t want it doesn’t specify it (could add a specific warning to the docs about bridges). 

> In addition, any data built with shortcuts would not work.

Specifying exclude_bridges/tunnels while we actually can’t guarantee it, is killing an upstream integration though. Thanks for pointing that out!

kevinkreiser commented:
in my opinion @dnesbitt61 is right, properly allowing the option has major drawbacks:

1. the routes are crap or cant be found
2. shortcuts have to be disabled since bridge and tunnel do not preclude a shortcut
3. the cost of disabling shortcuts and making big cuts in the graph at these edges is an operational nightmare (ie something you dont want to deal with if you are on call for a valhalla service with a latency metric hooked up to your pager haha)

that said i dont see why the main project cannot support such insanity optionally as @nilsnolde was alluding to. it would be possible to do a few things:

1. as part of the server configuration specify whether these options are to be allowed in a request
2. in src/worker.cc if it is detected that one of these options is set and its not allowed, return a 400
3. otherwise if one of these options is set and it is allowed, then disable shortcuts use (do we have a flag for that already, we should add that anyway also behind a server config since its not ideal by default)
4. default these dangerous options to off but allow individuals to turn them on with their configs

wouldn't such an approach make everyone happy? the novice wouldnt experience ill effects and the expert doing more offline computation with isochrones would be happy to configure their service to allow it and deal with not finding routes and extreme latency. what am i missing here? we already have a precedent for this with map matching where only some options can be modified at request time. this would be exactly the same approach.

azarz commented:
Thanks for the comments!

> In addition, any data built with shortcuts would not work.

Thank you for pointing that out, I didn't have that in mind at all.

> wouldn't such an approach make everyone happy?

It would make me happy for sure!
Are there benchmarks of the cost of disabling shortcuts? If it impacts performance too much, I think there should also be the possibility to allow those options as part of the server configuration even if shortcuts are enabled: the user could accept that those options are sometimes not respected on large routes/isochrones if the performance gain is significant. It would certainly be the case for us, as we run a web service with high sollicitation: we value performance over precision.

kevinkreiser commented:
> Are there benchmarks of the cost of disabling shortcuts

iirc they drop performance by half for the bidirectional a* algorithm. that said removing all bridges on all the highways that pass over smaller roads (ie exclude_bridges) is going to kill performance. i suspect it will fail the route and then do a second pass with a relaxed hierarchy limit (for routes of non trivial length). at any rate you can test the combination of both of these easily with a couple of hacks. just hack the costing models allowed function to disallow shortcuts and disallow bridges and see what the performance does compared to normal.

azarz commented:
> in src/worker.cc if it is detected that one of these options is set and its not allowed, return a 400

I'm starting to look into that, and I've run into several issues:
1. Which error code should we use? I thought it might be a 5xx because it seems to me that this is a Tyr project error, but I'm not sure
2. I've managed to access whether the parameter is set in the request in worker.cc, but I don't seem to find a way to access the server configuration in order to check if the parameter is allowed (which would be in a new option in the server configuration I suppose).

kevinkreiser commented:
the error code should be a 400 because its a request that invalid. as for the configuration you must add a value to the configuration generator (and then to all the default configs) here: https://github.com/valhalla/valhalla/blob/master/scripts/valhalla_build_config#L253 (plus a description lower down). then in loki_worker you must parse this config value so you then use it to throw a new error, probably you'll want to create the error somewhere around here: https://github.com/valhalla/valhalla/blob/master/src/worker.cc#L104 and you'd want to check the value  with the server configuration maybe somewhere in this function: https://github.com/valhalla/valhalla/blob/master/src/loki/worker.cc#L67

azarz commented:
> the error code should be a 400 because its a request thats invalid.
I'm sorry that was unclear, I was mentionning the valhalla error code (as in https://github.com/valhalla/valhalla/blob/master/docs/api/turn-by-turn/api-reference.md#internal-error-codes-and-conditions)

Thank you so much for you precise indications!

kevinkreiser commented:
Ah sorry yep add a new valhalla error code in the 100s

azarz commented:
> 1. as part of the server configuration specify whether these options are to be allowed in a request

Added new `service_limits.allow_hard_exclusions` option to server config.
https://github.com/valhalla/valhalla/pull/3733/files#diff-941fddb7a7fa73393545caf4c96cb1831af9d79831c4e082d9ee216d228a1a54R260
Defaults to `false`, considered `false` if not present.

> 2. in src/worker.cc if it is detected that one of these options is set and its not allowed, return a 400

https://github.com/valhalla/valhalla/pull/3733/files#diff-941fddb7a7fa73393545caf4c96cb1831af9d79831c4e082d9ee216d228a1a54R74
Added that check, the vahalla error code returned is `145 - Hard exclusions not allowed on this server`
https://github.com/valhalla/valhalla/pull/3733/files#diff-dda1eacb6be34d7946468f09844f48890c1266509f634990ec27263271bd6f48R84

> 3. otherwise if one of these options is set and it is allowed, then disable shortcuts use (do we have a flag for that already, we should add that anyway also behind a server config since its not ideal by default)

For the time being I have not done that. I haven't found a flag disabling shortcuts, and wasn't sure of all the implications of creating it. Moreover, I ran some tests on data I know, and the fact that bridges and tunnels aren't checked in the shortcut creation don't seem to be a big issue.

> 4. default these dangerous options to off but allow individuals to turn them on with their configs

https://github.com/valhalla/valhalla/pull/3733/files#diff-1117ccfb8150146a5d28a6cbc7ae7a8ca059a64975a54c31a8461d2b44a835d3R134
Options off by default.

nilsnolde commented on test/gurka/test_exclude_bridges.cc:
```diff
@@ -0,0 +1,136 @@
+
+TEST_F(ExcludeBridgesTest, BridgesInTheBeginning) {
+  // Without options
+  for (const auto& costing : kSupportedCostingModels) {
```
You could probably even be able to come up with a smarter parameterization so you could just write one test which is parameterized with both the costing model and the `exclude_*` request option as pairs. Then we'd almost save 2/3 of the test code. Though not sure if Valhalla allows a road to be tagged with both `bridge` and `tunnel` :smile: at least `tolls` can be tagged on the same way in any case so you might get away with one map to rule them all. It's mostly about saving quite a bit of code. The testing is really thorough, thanks for that!

azarz replied:
Thank you for the suggestion! I will look into that

nilsnolde reviewed (changes requested):
@azarz thanks for the PR! There's one quick thing about reverse access and a suggestion for less verbose testing. And what do you think about extending or limiting the supported list of costing models?

nilsnolde commented:
I also think it should disable using shortcuts. There's tons of bridges on motorways which will be contracted away. An `exclude_bridges` that doesn't exclude all bridges is a bit problematic I find. Not sure how to pass that into the A* for every request..

kevinkreiser commented:
You could do the shortcut checking in costing in the allowed function possibly

azarz commented on docs/api/turn-by-turn/api-reference.md:
```diff
@@ -128,6 +128,13 @@ The following options are available for `auto`, `bus`, `taxi`, and `truck` costi
 | `include_hov3` | A boolean value which indicates the desire to include HOV roads with a 3-occupant requirement in the route when advantageous. Default false. |
 | `include_hot` | A boolean value which indicates the desire to include tolled HOV roads which require the driver to pay a toll if the occupant requirement isn't met. Default false. |
 
+The following options are available for `auto`, `bus`, `taxi`, and `truck` and `pedestrian` costing methods.
```
Yes you're right about all costing models. I'll add the option in all of them.

azarz commented on src/sif/pedestriancost.cc:
```diff
@@ -669,7 +672,9 @@ bool PedestrianCost::AllowedReverse(const baldr::DirectedEdge* edge,
       //      (opp_edge->max_up_slope() > max_grade_ || opp_edge->max_down_slope() > max_grade_) ||
       opp_edge->use() == Use::kTransitConnection || opp_edge->use() == Use::kEgressConnection ||
-      opp_edge->use() == Use::kPlatformConnection) {
+      opp_edge->use() == Use::kPlatformConnection || (exclude_bridges_ && !pred.bridge() && edge->bridge()) ||
+      (exclude_tunnels_ && !pred.tunnel() && edge->tunnel()) ||
+      (exclude_tolls_ && !pred.toll() && edge->toll())) {
```
Thanks a lot! Nice find

azarz commented on src/sif/autocost.cc:
```diff
@@ -467,7 +470,10 @@ bool AutoCost::AllowedReverse(const baldr::DirectedEdge* edge,
       (!allow_destination_only_ && !pred.destonly() && opp_edge->destonly()) ||
       (pred.closure_pruning() && IsClosed(opp_edge, tile)) ||
-      (exclude_unpaved_ && !pred.unpaved() && opp_edge->unpaved()) || !IsHOVAllowed(opp_edge)) {
+      (exclude_unpaved_ && !pred.unpaved() && edge->unpaved()) || !IsHOVAllowed(edge) ||
```
Thanks, I missed that

azarz commented:
I'm trying to reduce the code used for testing the features, using the gtest `Combine` method, however I'm getting a compilation error that I shouldn't, according to the `Combine` [documentation](http://google.github.io/googletest/reference/testing.html#INSTANTIATE_TEST_SUITE_P) and [examples](https://github.com/google/googletest/blob/main/googletest/samples/sample8_unittest.cc)...

The code :
```c++
#include "gurka.h"

#include <gtest/gtest.h>

using namespace valhalla;

namespace {
  const std::vector<std::string> kSupportedCostingModels = {
    "auto",
    "taxi",
    "bus",
    "truck",
    "pedestrian",
    "bicycle",
    "motor_scooter",
    "motorcycle"
  };

  const std::vector<std::string> kExclusionParameters = {
    "exclude_bridges",
    "exclude_tunnels",
    "exclude_tolls"
  };
} // namespace

class ExclusionTest : public ::testing::TestWithParam<std::vector<std::string>> {
protected:
  static gurka::map map;

  static void SetUpTestSuite() {
    constexpr double grid_size_meters = 100.;

    const std::string ascii_map = R"(
   E----F----G----H----I----A----J----K----L
                                 |    |
                                 |    |
                                 |    |
                                 |    |
                                 M----N
    )";

    const gurka::ways ways = {
        {"EF", {{"highway", "residential"}, {"bridge", "yes"}, {"tunnel", "yes"}, {"toll", "yes"}}},
        {"FG", {{"highway", "residential"}, {"bridge", "yes"}, {"tunnel", "yes"}, {"toll", "yes"}}},
        {"GH", {{"highway", "residential"}}},
        {"HI", {{"highway", "residential"}, {"bridge", "yes"}, {"tunnel", "yes"}, {"toll", "yes"}}},
        {"IA", {{"highway", "residential"}}},
        {"IJ", {{"highway", "residential"}}},
        {"JK", {{"highway", "residential"}, {"bridge", "yes"}, {"tunnel", "yes"}, {"toll", "yes"}}},
        {"KL", {{"highway", "residential"}}},
        {"JM", {{"highway", "residential"}}},
        {"MN", {{"highway", "residential"}}},
        {"NK", {{"highway", "residential"}}},
    };

    const auto layout = gurka::detail::map_to_coordinates(ascii_map, grid_size_meters);
    map = gurka::buildtiles(layout, ways, {}, {}, "test/data/hard_exclude",
                            {{"service_limits.allow_hard_exclusions", "true"}});
  }

  void check_result(const std::string& exclude_parameter_value,
                    const std::vector<std::string>& waypoints,
                    const std::vector<std::string>& expected_names,
                    const std::vector<std::string>& props = {}) {

    const std::string& costing = props[0];
    const std::string& exclude_parameter = props[1];
    const auto result = gurka::do_action(valhalla::Options::route, map, waypoints, costing,
                                          {{"/costing_options/" + costing + "/" + exclude_parameter, exclude_parameter_value}});
    gurka::assert::raw::expect_path(result, expected_names);
  }

};

gurka::map ExclusionTest::map = {};

TEST_P(ExclusionTest, InTheMiddle) {
  check_result("0", {"I","L"}, {"IJ", "JK", "KL"}, GetParam());
  check_result("1", {"I","L"}, {"IJ", "JM", "MN", "NK", "KL"}, GetParam());
}

TEST_P(ExclusionTest, InTheBeginning) {
  check_result("0", {"E", "H"}, {"EF", "FG", "GH"}, GetParam());
  check_result("1", {"E", "H"}, {"EF", "FG", "GH"}, GetParam());
}

TEST_P(ExclusionTest, InTheEnd) {
  check_result("0", {"H", "E"}, {"GH", "FG", "EF"}, GetParam());
  check_result("1", {"H", "E"}, {"GH", "FG", "EF"}, GetParam());
}

INSTANTIATE_TEST_SUITE_P(ExcludePropsTest,
                         ExclusionTest,
                         ::testing::Combine(kSupportedCostingModels, kExclusionParameters));

``` 

The error :
> error: could not convert ‘testing::Combine(const Generator& ...) [with Generator = {std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >}]({anonymous}::kExclusionParameters)’ from ‘testing::internal::CartesianProductHolder<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >’ to ‘testing::internal::ParamGenerator<std::vector<std::__cxx11::basic_string<char> > >’
   94 |                          ::testing::Combine(kSupportedCostingModels, kExclusionParameters));
      |                          ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                            |
      |                                            testing::internal::CartesianProductHolder<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >

nilsnolde commented:
Great, looks almost done @azarz , and the test code is much more concise.

Not sure how `Combine` exactly is supposed to work but seems it has trouble with vectors as input, I think the test suite always expect one of its value wrappers as input.. try wrapping each vector in `::testing::ValuesIn` when passing it to `Combine`. Or do it as other test files are doing it, e.g. `test_expansion.cc` or `test_via_waypoints.cc`.

nilsnolde commented on src/sif/truckcost.cc:
```diff
@@ -413,7 +417,11 @@ bool TruckCost::AllowedReverse(const baldr::DirectedEdge* edge,
+      (exclude_bridges_ && !pred.bridge() && opp_edge->bridge()) ||
+      (exclude_tunnels_ && !pred.tunnel() && opp_edge->tunnel()) ||
+      (exclude_tolls_ && !pred.toll() && opp_edge->toll()) ||
+      (edge->is_shortcut() && (exclude_bridges_ || exclude_tunnels_))) {
```
copy/paste got ya again :smile: (same in others)

azarz commented:
I get a similar compilation error using no standard library vectors.
```c++
INSTANTIATE_TEST_SUITE_P(ExcludePropsTest,
                         ExclusionTest,
                         ::testing::Combine(::testing::Values("auto",
                                                              "taxi",
                                                              "bus",
                                                              "truck",
                                                              "pedestrian",
                                                              "bicycle",
                                                              "motor_scooter",
                                                              "motorcycle"
                                                              ),
                                            ::testing::Values("exclude_bridges",
                                                              "exclude_tunnels",
                                                              "exclude_tolls"
                                                              )
                                            )
                          );
```

I should maybe use a different way to do the cartesian product between the two arrays

nilsnolde commented:
Sorry, should've provided a code suggestion instead of describing it.. I think this should work:
```cpp
INSTANTIATE_TEST_SUITE_P(ExcludePropsTest,
                         ExclusionTest,
                         ::testing::Combine(
                                            ::testing::ValuesIn(kSupportedCostingModels),
                                            ::testing::ValuesIn(kExclusionParameters)
                                            )
                          );
```

If it doesn't, please see the examples here: 
- https://github.com/valhalla/valhalla/blob/52f869ea2cc192ab31e9d5e75170cab29694059c/test/gurka/test_expansion.cc#L124
- https://github.com/valhalla/valhalla/blob/52f869ea2cc192ab31e9d5e75170cab29694059c/test/gurka/test_via_waypoints.cc#L223

nilsnolde commented:
One last small optimization could be to put it all in one test file. You can then put the map in the anonymous namespace and reuse the vectors, saves some duplication and one file:)

azarz commented:
I've updated the tests, now there is only one file that tests the feature both when the exclusions are allowed and not allowed on the server.

nilsnolde reviewed (approved):
Thanks, looks perfect for me!

nilsnolde commented:
There are some conflicts but you can wait until someone else approved before resolving them, maybe another PR will be merged before.

kevinkreiser commented:
@azarz we want to get this merged some time soonish, and get it reviewed but with a pr like this there are lots of potential for regressions, both on results and on performance. what our goal is here, is to allow your feature to make it in, but in the case a user doesnt opt into using it performance is vastly unchanged and the results are identical to previous. this means we have to run some scripting with some corpus of requests and compare the results. it will take us a bit of time to do it and report back our findings but we'll get to it in time. apologies for the wait and thank you for your contribution.

azarz commented:
No worries! It's only natural to do thorough testing when performance could be impacted :wink:

jmkerloch commented:
We are currently using this branch for the development of the new version of IGN geoportail. This is working as expected but we would prefer to use branch from valhalla repository.

@kevinkreiser do you have an idea of when this PR could be merged ? How could we help the review process ?

azarz commented:
I have rebased the branch on valhalla's master in order to resolve the previous conflicts.

ADMUKHE commented:
@nilsnolde @kevinkreiser we can help with comparing the test results and see if there is a difference incase the user dosent opt to use it,  to speed up the  this PR.  What should the approach be? We work on a seperate branch?

azarz commented:
I have rebased the branch on valhalla's master.

ADMUKHE commented:
@nilsnolde in this case if we need shortcuts, can we also extend this PR to add tunnel attribute in shortcut creation [](https://github.com/valhalla/valhalla/blob/3806832a11fe250e08f6f21c9f55da0e052bfc80/src/mjolnir/shortcutbuilder.cc#L69C1-L71C28), but for exclude_bridge specifically we need to make sure shortcut is disabled

azarz commented:
Closed since https://github.com/valhalla/valhalla/pull/4524 is merged and includes the changes of this PR
